master/buildbot/process/buildrequestmerger.py

```diff
@@ -0,0 +1,248 @@
+
+        # For every builderName in this buildset, check which ones can be merged
+        breqsToMerge = {}
+        for builderName in builderNames:
+            builderMergeStart = time.time()
+            mergeBrid = self._getMergeBrid(
+                self.master.botmaster.builders[builderName], builderName,
+                priority, codebase, branch, revision, properties)
+            if mergeBrid:
+                breqsToMerge[builderName] = mergeBrid
+
+            buildsetLog[builderName] = {
+                'elapsed': time.time() - builderMergeStart,
+                'mergeBrid': mergeBrid
+            }
+
+        buildsetLog['elapsed'] = time.time() - start
+
+        log.msg(json.dumps(buildsetLog))
+
+        return self.master.db.buildsets.addBuildset(
+            sourcestampsetid=sourcestampsetid,
+            reason=reason,
+            properties=properties,
+            triggeredbybrid=triggeredbybrid,
+            builderNames=builderNames,
+            breqsToMerge=breqsToMerge,
+            external_idstring=external_idstring,
+            _reactor=_reactor, )
+
+    def _getSourceStampInfo(self, sourcestampsetid):
+        """
+        :param str sourcestampsetid:
+        :return tuple(str,str,str):
+            (codebase, branch, revision) for the given `sourcestampsetid`
+        """
+        ss_tbl = self.master.db.model.sourcestamps
+
+        def __getSourcestampInfo(conn):
+            res = conn.execute(
+                sa.select([
+                    ss_tbl.c.branch,
+                    ss_tbl.c.revision,
+                    ss_tbl.c.repository,
+                    ss_tbl.c.codebase,
+                ]).where(ss_tbl.c.sourcestampsetid == sourcestampsetid))
+
+            row = res.fetchone()
+            return row.codebase, row.branch, row.revision
+
+        return self.master.db.pool.do(__getSourcestampInfo)
+
+    def _getMergeBrid(self, builder, builderName, priority, codebase, branch,
+                      revision, properties):
+        """
+        :param Builder builder:
+        :param str builderName:
+        :param str codebase:
+        :param int priority:
+        :param str branch:
+        :param str revision:
+        :param dict(str,str) properties:
+
+        :return str or None:
+            Build request id for a request that can be merged into (matches
+            buiderName, properties and sourcestamp information).
+
+            `None` if no match was found.
+        """
+        if 'selected_slave' in properties:
+            return None
+
+        # Get an initial list of all breqs that match the given source stamp data
+        matchingBrids = self._getBuildRequestIdsMatchingSourceStamp(
+            builderName, priority, codebase, branch, revision)
+
+        # Check if relevant properties match
+        for otherBuildsetid, otherBrid in matchingBrids:
+            otherProperties = self.properties_cache.get(otherBuildsetid)
```

Comment:
Would this not result in a possible N+1 select if the properties are not in the cache?

Reply:
Yep, one query for each `buildsetid` missing from the cache.

I'll make this smarter and use `IN <id_list>`